File: zig/isogram/isogram.zig
pub fn isIsogram(str: []const u8) bool {
    var seen: u26 = 0;

    for (str) |c| {
        const bit: u8 = switch (c) {
            'A'...'Z' => c - 'A',
            'a'...'z' => c - 'a',
            else => continue,
        };
        const mask = @as(u26, 1) << @intCast(bit);
        if ((seen & mask) != 0) return false;
        seen |= mask;
    }

    return true;
}

File: c/isogram/isogram.c
#include "isogram.h"
#include <ctype.h>
#include <stdbool.h>

bool is_isogram(const char phrase[]) {
  if (!phrase) {
    return false;
  }

  unsigned int seen = 0;

  for (const char *c = phrase; *c != '\0'; ++c) {
    if (!isalpha(*c)) {
      continue;
    }

    unsigned char bit = tolower(*c) - 'a';
    if (bit >= 26) {
      return false;
    }

    unsigned int mask = 1u << bit;
    if (seen & mask) {
      return false;
    }
    seen |= mask;
  }

  return true;
}
